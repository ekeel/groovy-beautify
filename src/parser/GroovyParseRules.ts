import ParseRule from "./ParseRule";

export default [
  new ParseRule("blockcomment", { start: "/*", end: "*/", exclusive: true }),
  new ParseRule("linecomment", { start: "//", end: /(?=\n)/, exclusive: true }),
  new ParseRule("multilinestring", { start: "'''", end: "'''", exclusive: true }),
  new ParseRule("multilinestring", { start: '"""', end: '"""', exclusive: true }),
  new ParseRule("string", { start: '"', end: '"', skip: '\\"', exclusive: true }),
  new ParseRule("string", { start: "'", end: "'", skip: "\\'", exclusive: true }),
  new ParseRule("regexp", { start: /\//g, end: /\//g, skip: /\'/g, exclusive: true }),
  new ParseRule("numeric", { start: /0[xX][0-9a-fA-F]+\b/g }),
  new ParseRule("numeric", { start: /[+-]?\d+(?:(?:\.\d*)?(?:[eE][+-]?\d+)?)?\b/g }),
  new ParseRule("boolean", { start: /(?:true|false)\b/g }),
  new ParseRule("block", { start: "{", end: "}" }),
  new ParseRule("square", { start: ["?[", "["], end: "]" }),
  new ParseRule("round", { start: "(", end: ")" }),
  new ParseRule("dot", { start: /((\?|\*|\&)+)?\.(?=[^\d])(\*)?/g }), // Includes dot, spread dot, method reference, safe dot, safe chain dot and import all
  new ParseRule("delimiters", { start: [",", ":", ";"] }),
  new ParseRule("incdec", { start: ["--", "++", "**"] }),
  new ParseRule("negation", { start: "!" }),
  // prettier-ignore
  new ParseRule("operators", { start: ['--', '-=', '->', '-', '::', '!==', '!=', '?:', '?=', '?', '..<', '..', '**=', '**', 
      '*=', '*', '/=', '/', '&&', '&=', '&', '%=', '%', '^=', '^', '++', '+=', '+', '<..<', '<..', '<<=', '<<', '<=>', '<=', '<', 
      '===', '==~', '==', '=~', '=', '>=', '>>=', '>>>=', '>>', '>', '|=', '||', '|', '~'] }),
  // prettier-ignore
  new ParseRule("keywords", {
    start: ['abstract', 'assert', 'boolean', 'break', 'byte', 'case', 'catch', 'char', 'class', 'continue', 
        'default', 'def', 'double', 'do', 'else', 'enum', 'extends', 'finally', 'final', 'float', 'for', 'goto', 'if', 
        'implements', 'import', 'instanceof', 'interface', 'int', 'long', 'native', 'new', 'package', 'private', 'protected', 
        'public', 'return', 'short', 'static', 'strictfp', 'super', 'switch', 'synchronized', 'this', 'throws', 'throw', 
        'transient', 'try', 'void', 'volatile', 'while', 'with'],
    wholeword: true,
  }),
  new ParseRule("identifiers", { start: /[a-zA-Z_$][a-zA-Z0-9_$]*\b/g, wholeword: true }),
];
