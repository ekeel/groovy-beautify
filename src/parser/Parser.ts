import ParseRule, { MatchExpression } from "./ParseRule";

export type CodeBlock = {
  type: string;
  start: string;
  end?: string;
  children?: Array<CodeBlock>;
};

export class Parser {
  text: string;
  textLength: number;
  textPosition: number;
  rules: ParseRule[];

  /**
   * @param text - Text to parse
   * @param rules - Array of parsing rules. If none of specified rules match default rule with type text will be used.
   */
  constructor(text: string, rules?: ParseRule[]) {
    this.text = text;
    this.textLength = text ? text.length : 0;
    this.textPosition = 0;
    this.rules = rules || [];
  }

  scan() {
    if (this.textPosition >= this.textLength) return null;

    return this.text[this.textPosition++];
  }

  isDone() {
    return this.textPosition >= this.textLength;
  }

  move(length: number) {
    this.textPosition += length;
  }

  match(expression?: MatchExpression | MatchExpression[], wholeword?: boolean): string | undefined {
    let result;
    if (expression instanceof RegExp) {
      const modifiedExpression = new RegExp("^(" + expression.source + ")", expression.flags);
      const text = this.text.substring(this.textPosition).match(modifiedExpression)?.[0];
      if (text?.length) {
        result = text;
      }
    } else if (expression instanceof Array) {
      for (const exp of expression) {
        let match = this.match(exp);
        if (match) {
          result = match;
          break;
        }
      }
    } else if (typeof expression === "string") {
      if (this.text.substring(this.textPosition).startsWith(expression)) {
        result = expression;
      }
    }
    if (result && wholeword) {
      const prevSymbol = this.text.charAt(this.textPosition - 1);
      const nextSymbol = this.text.charAt(this.textPosition + result.length);
      if (prevSymbol.match(/[A-Za-z0-9_]/) || nextSymbol.match(/[A-Za-z0-9_]/)) {
        return undefined;
      }
    }
    return result;
  }

  matchStart(rule: ParseRule) {
    return this.match(rule.start, rule.wholeword);
  }
  matchEnd(rule: ParseRule) {
    return this.match(rule.end, rule.wholeword);
  }
  matchSkip(rule: ParseRule) {
    return this.match(rule.skip, rule.wholeword);
  }

  createDefaultTextObject(text: string) {
    return {
      type: "text",
      start: text,
    };
  }

  parse() {
    const root: CodeBlock = { type: "root", start: "" };
    const ruleStack = [] as Array<ParseRule>;
    const objStack = [root];

    while (!this.isDone()) {
      const activeRule = ruleStack[ruleStack.length - 1];
      const activeObj = objStack[objStack.length - 1];
      const activeObjLastChild = activeObj.children?.[activeObj.children.length - 1];

      if (activeRule) {
        // Check skip first
        if (activeRule.skip) {
          const activeRuleSkipText = this.matchSkip(activeRule);
          if (activeRuleSkipText) {
            // Append to existing text block or create a new one
            if (activeObjLastChild && activeObjLastChild.type === "text") {
              activeObjLastChild.start += activeRuleSkipText;
            } else {
              if (!activeObj.children) {
                activeObj.children = [];
              }
              activeObj.children.push(this.createDefaultTextObject(activeRuleSkipText));
            }
            this.move(activeRuleSkipText.length);
            continue;
          }
        }
        if (activeRule.end) {
          // Check rule end
          const activeRuleText = this.matchEnd(activeRule);
          if (activeRuleText) {
            activeObj.end = activeRuleText;
            ruleStack.pop();
            objStack.pop();
            this.move(activeRuleText.length);
            continue;
          }
        }
      }
      // Check rule start
      if (!activeRule?.exclusive) {
        const matchedRule = this.rules.find((r) => this.matchStart(r));
        if (matchedRule) {
          const matchedRuleText = this.matchStart(matchedRule);
          if (matchedRuleText) {
            const newObj = {
              type: matchedRule.name,
              start: matchedRuleText,
            };
            if (!activeObj.children) {
              activeObj.children = [];
            }
            activeObj.children.push(newObj);
            // If we need to find matched rule
            if (matchedRule.end) {
              ruleStack.push(matchedRule);
              objStack.push(newObj);
            }
            this.move(matchedRuleText.length);
            continue;
          }
        }
      }

      // Read char by char
      const ch = this.scan();
      if (ch) {
        if (activeObjLastChild && activeObjLastChild.type === "text") {
          activeObjLastChild.start += ch;
        } else {
          if (!activeObj.children) {
            activeObj.children = [];
          }
          activeObj.children.push(this.createDefaultTextObject(ch));
        }
      }
    }

    return root;
  }
}
