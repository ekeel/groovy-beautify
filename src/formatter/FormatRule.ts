import { CodeBlock } from "../parser/Parser";
import { Formatter } from "./Formatter";

export default class FormatRule {
  formatter: Formatter;

  constructor(formatter: Formatter) {
    this.formatter = formatter;
  }

  matches(cb: CodeBlock): boolean {
    return true;
  }
  /* Modifies next sibling text before adding it */
  afterSelf(nextText: string, indent: number): string {
    return nextText;
  }
  /* Modifies previous sibling text before adding it */
  beforeSelf(prevText: string, indent: number): string {
    return prevText;
  }
  /* Modifies child text before adding it */
  beforeChild(childText: string, indent: number): string {
    return childText;
  }

  formatStart(cb: CodeBlock, indent: number): string {
    return cb.start ?? "";
  }

  formatEnd(cb: CodeBlock, indent: number): string {
    return cb.end ?? "";
  }

  allowBreak(cb: CodeBlock) {
    return false;
  }

  formatChildren(parent: CodeBlock, indent: number) {
    return (
      parent?.children?.reduce((res: string, child, i, children) => {
        const childFormatRule = this.formatter.rules.find((r) => r.matches(child));
        const parentFormatRule = this.formatter.rules.find((r) => r.matches(parent));
        const prevFormatRule = this.formatter.rules.find((r) => r.matches(children[i - 1]));
        const nextFormatRule = this.formatter.rules.find((r) => r.matches(children[i + 1]));

        let childString = this.formatter.format(child, indent);

        if (prevFormatRule) {
          childString = prevFormatRule.afterSelf(childString, indent);
        }
        if (nextFormatRule) {
          childString = nextFormatRule.beforeSelf(childString, indent);
        }
        if (parentFormatRule) {
          childString = parentFormatRule.beforeChild(childString, indent);
        }

        const lastLineLength = res.split("\n").at(-1)?.length ?? 0;
        const childFisrtLineLength = childString.split("\n").at(0)?.length ?? 0;
        if (lastLineLength + childFisrtLineLength > this.formatter.options.width && childFormatRule?.allowBreak(child)) {
          childString = "\n" + childString.trimStart();
          if (parentFormatRule) {
            childString = parentFormatRule.beforeChild(childString, indent + 1);
          }
        }
        res += childString;

        return res;
      }, "") || ""
    );
  }
}
