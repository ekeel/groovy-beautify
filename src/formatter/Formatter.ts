import { CodeBlock } from "../parser/types";
import FormatRule from "./FormatRule";

export type FormatterOptions = {
  width: number;
};

export class Formatter {
  rules: FormatRule[];
  options: FormatterOptions;

  constructor(rules?: Array<typeof FormatRule>, options?: FormatterOptions) {
    this.rules = rules?.map((r) => new r(this)) || [];
    this.options = options || { width: 80 };
  }

  format(obj: CodeBlock, indent = 0): string {
    let text = "";

    if (obj) {
      const formatRule = this.rules.find((r) => r.matches(obj));
      if (obj.start !== undefined) {
        if (formatRule) {
          text += formatRule.formatStart(obj, indent);
        }
      }

      if (formatRule) {
        text += formatRule.formatChildren(obj, indent);
      }

      if (obj.end !== undefined) {
        if (formatRule) {
          text += formatRule.formatEnd(obj, indent);
        }
      }
    }
    return text;
  }
}
