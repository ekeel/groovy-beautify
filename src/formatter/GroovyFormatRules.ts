import { CodeBlock } from "../parser/Parser";
import FormatRule from "./FormatRule";

class BaseBlockRule extends FormatRule {
  beforeChild(childText: string, indent: number): string {
    let text = childText;
    const trimmedLeft = trimSpacesAndTabsLeft(text);
    if (trimmedLeft.startsWith("\n")) {
      text = "\n" + padLeft(trimmedLeft.trimStart(), indent);
    }
    const trimmedRight = trimSpacesAndTabsRight(text);
    if (trimmedRight.endsWith("\n")) {
      text = padRight(trimmedRight, indent);
    }
    return text;
  }
}

class BlockFormatRule extends BaseBlockRule {
  matches(cb: CodeBlock) {
    return cb?.type === "block";
  }

  beforeSelf(prevText: string) {
    const trimmedText = trimSpacesAndTabsRight(prevText);
    if (trimmedText.endsWith("\n")) {
      return trimmedText;
    } else {
      return prevText.trimEnd() + " ";
    }
  }

  formatEnd(cb: CodeBlock, indent: number) {
    return cb.end ? padLeft(cb.end, indent) : "";
  }

  formatChildren(cb: CodeBlock, indent: number) {
    let blockText = super.formatChildren(cb, indent + 1);
    blockText = blockText.trim();
    return "\n" + padLeft(blockText, indent + 1) + "\n";
  }
}

class InlineBlockFormatRule extends BaseBlockRule {
  matches(cb: CodeBlock) {
    return cb?.type === "round" || cb?.type === "square";
  }

  isMultiline(cb: CodeBlock) {
    return cb.children?.some((child) => child.start?.includes("\n"));
  }

  beforeSelf(prevText: string) {
    const trimmedText = trimSpacesAndTabsRight(prevText);
    if (trimmedText.endsWith("\n")) {
      return trimmedText;
    } else {
      return prevText;
    }
  }

  formatEnd(cb: CodeBlock, indent: number) {
    if (this.isMultiline(cb)) {
      return cb.end ? padLeft(cb.end, indent) : "";
    } else {
      return cb.end ?? "";
    }
  }

  formatChildren(cb: CodeBlock, indent: number) {
    if (this.isMultiline(cb)) {
      let blockText = super.formatChildren(cb, indent + 1);
      blockText = blockText.trim();
      return "\n" + padLeft(blockText, indent + 1) + "\n";
    } else {
      let blockText = super.formatChildren(cb, indent);
      return blockText.trim();
    }
  }
}

class DotSyntaxFormatRule extends FormatRule {
  matches(cb: CodeBlock) {
    return cb?.type === "dot";
  }
  beforeSelf(prevText: string, indent: number): string {
    const trimmedText = trimSpacesAndTabsRight(prevText);
    if (trimmedText.endsWith("\n")) {
      return padRight(trimmedText, indent + 1);
    } else {
      return prevText;
    }
  }
}

class OperatorsRule extends FormatRule {
  matches(cb: CodeBlock) {
    return cb?.type === "operators";
  }

  beforeSelf(prevText: string): string {
    return trimSpacesAndTabsRight(prevText) + " ";
  }

  afterSelf(nextText: string): string {
    return " " + nextText.trimStart();
  }

  allowBreak(): boolean {
    return true;
  }
}

class DelimitersRule extends FormatRule {
  matches(cb: CodeBlock) {
    return cb?.type === "delimiters";
  }

  beforeSelf(prevText: string): string {
    return trimSpacesAndTabsRight(prevText);
  }

  afterSelf(nextText: string): string {
    return " " + trimSpacesAndTabsLeft(nextText);
  }
}

class KeywordRule extends FormatRule {
  matches(cb: CodeBlock) {
    return cb?.type === "keywords";
  }

  beforeSelf(prevText: string): string {
    const trimmedText = trimSpacesAndTabsRight(prevText);
    if (trimmedText.endsWith("\n")) {
      return trimmedText;
    } else {
      return trimmedText + " ";
    }
  }

  afterSelf(nextText: string): string {
    return " " + nextText.trimStart();
  }

  allowBreak(): boolean {
    return true;
  }
}

function padLeft(text: string, indent: number) {
  return "".padStart(indent * 4) + text;
}
function padRight(text: string, indent: number) {
  return text + "".padStart(indent * 4);
}

function trimSpacesAndTabsLeft(text: string) {
  return text.replace(/^( |\t)+/, "");
}
function trimSpacesAndTabsRight(text: string) {
  return text.replace(/( |\t)+$/, "");
}

export default [
  BlockFormatRule,
  InlineBlockFormatRule,
  DotSyntaxFormatRule,
  KeywordRule,
  OperatorsRule,
  DelimitersRule,
  FormatRule,
];
