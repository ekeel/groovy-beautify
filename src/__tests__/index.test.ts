import { Parser } from "../parser/Parser";
import GroovyParseRules from "../parser/GroovyParseRules";
import { Formatter } from "../formatter/Formatter";
import GroovyFormatRules from "../formatter/GroovyFormatRules";

const TEST_SCRIPT = `
  def bumpVersion(String target,    String version_type, Boolean reset =   false) {    def
	versionMap =
['major':0, 'minor' : 1, 'patch':   2]
def versionArray = target.findAll(/\d+\.\d+\.\d+/)[0].tokenize('.')
try
{        def   index =     versionMap.get(version_type);
	versionArray[index] =versionArray[index].toInteger() + 1
	if(   reset )
	{
		for(int i=2;i>index;     i--) {
			versionArray[i]    =    0            }        }
} catch(       Exception e) {        println("Unrecognized version type \\"version_type\\" (should be major, minor or patch)")    }
return             versionArray.join(                   '.'                       )
}

println(bumpVersion('1.2.3', 'minor', true))
`;

test("Parsing and formatting", () => {
  const parser = new Parser(TEST_SCRIPT, GroovyParseRules);
  const parsingResult = parser.parse();
  expect(parsingResult).toMatchSnapshot();

  const formatter = new Formatter(GroovyFormatRules);
  const formattingResult = formatter.format(parsingResult);
  expect(formattingResult).toMatchSnapshot();
});

export {};
