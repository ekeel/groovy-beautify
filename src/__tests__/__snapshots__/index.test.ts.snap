// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line comment 1`] = `
"{
    def shape = PathROIToolsAwt.getShape(roi) // Comment
    shape2 = transform.createTransformedShape(shape)
}"
`;

exports[`Long string with operators 1`] = `
"if (true) {
    def a = 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 
        + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 
        + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 
        + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 
        + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 
        + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 
        + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 
        + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 
        + 1 + 1 + 1 + 1 + 1 + 1 + 1;
}"
`;

exports[`Multiline Strings 1`] = `
Object {
  "children": Array [
    Object {
      "start": "
",
      "type": "text",
    },
    Object {
      "start": "if",
      "type": "keywords",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": "true",
          "type": "identifiers",
        },
      ],
      "end": ")",
      "start": "(",
      "type": "round",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": "
",
          "type": "text",
        },
        Object {
          "start": "def",
          "type": "keywords",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "string1",
          "type": "identifiers",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "=",
          "type": "operators",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "children": Array [
            Object {
              "start": "Some
multiline
String",
              "type": "text",
            },
          ],
          "end": "'''",
          "start": "'''",
          "type": "multilinestring",
        },
        Object {
          "start": "
",
          "type": "text",
        },
        Object {
          "start": "def",
          "type": "keywords",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "string2",
          "type": "identifiers",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "=",
          "type": "operators",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "children": Array [
            Object {
              "start": "Some
multiline
String",
              "type": "text",
            },
          ],
          "end": "\\"\\"\\"",
          "start": "\\"\\"\\"",
          "type": "multilinestring",
        },
        Object {
          "start": "
",
          "type": "text",
        },
      ],
      "end": "}",
      "start": "{",
      "type": "block",
    },
  ],
  "start": "",
  "type": "root",
}
`;

exports[`Multiline Strings 2`] = `
"
if (true) {
    def string1 = '''Some
multiline
String'''
    def string2 = \\"\\"\\"Some
multiline
String\\"\\"\\"
}"
`;

exports[`Multiline in square 1`] = `
"
def matrix = [
    - 0.998, - 0.070, 127256.994,
    0.070, - 0.998, 72627.371
]
   "
`;

exports[`Nested round 1`] = `
"
if (pathObject.hasChildren()) {
    newObject.addPathObjects(pathObject.getChildObjects().collect({
        transformObject(it, transform)
    }))
}
   "
`;

exports[`Parsing and formatting 1`] = `
Object {
  "children": Array [
    Object {
      "start": "def",
      "type": "keywords",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "start": "bumpVersion",
      "type": "identifiers",
    },
    Object {
      "children": Array [
        Object {
          "start": "String",
          "type": "identifiers",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "target",
          "type": "identifiers",
        },
        Object {
          "start": ",",
          "type": "delimiters",
        },
        Object {
          "start": "    ",
          "type": "text",
        },
        Object {
          "start": "String",
          "type": "identifiers",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "version_type",
          "type": "identifiers",
        },
        Object {
          "start": ",",
          "type": "delimiters",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "Boolean",
          "type": "identifiers",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "reset",
          "type": "identifiers",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "=",
          "type": "operators",
        },
        Object {
          "start": "   ",
          "type": "text",
        },
        Object {
          "start": "false",
          "type": "identifiers",
        },
      ],
      "end": ")",
      "start": "(",
      "type": "round",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": "    ",
          "type": "text",
        },
        Object {
          "start": "def",
          "type": "keywords",
        },
        Object {
          "start": "
",
          "type": "text",
        },
        Object {
          "start": "versionMap",
          "type": "identifiers",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "=",
          "type": "operators",
        },
        Object {
          "start": "
",
          "type": "text",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "start": "major",
                  "type": "text",
                },
              ],
              "end": "'",
              "start": "'",
              "type": "string",
            },
            Object {
              "start": ":",
              "type": "delimiters",
            },
            Object {
              "start": "0",
              "type": "text",
            },
            Object {
              "start": ",",
              "type": "delimiters",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "children": Array [
                Object {
                  "start": "minor",
                  "type": "text",
                },
              ],
              "end": "'",
              "start": "'",
              "type": "string",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "start": ":",
              "type": "delimiters",
            },
            Object {
              "start": " 1",
              "type": "text",
            },
            Object {
              "start": ",",
              "type": "delimiters",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "children": Array [
                Object {
                  "start": "patch",
                  "type": "text",
                },
              ],
              "end": "'",
              "start": "'",
              "type": "string",
            },
            Object {
              "start": ":",
              "type": "delimiters",
            },
            Object {
              "start": "   2",
              "type": "text",
            },
          ],
          "end": "]",
          "start": "[",
          "type": "square",
        },
        Object {
          "start": "
            ",
          "type": "text",
        },
        Object {
          "start": "def",
          "type": "keywords",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "versionArray",
          "type": "identifiers",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "=",
          "type": "operators",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "target",
          "type": "identifiers",
        },
        Object {
          "start": ".",
          "type": "dot",
        },
        Object {
          "start": "findAll",
          "type": "identifiers",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "start": "d+.d+.d+",
                  "type": "text",
                },
              ],
              "end": "/",
              "start": "/",
              "type": "regexp",
            },
          ],
          "end": ")",
          "start": "(",
          "type": "round",
        },
        Object {
          "children": Array [
            Object {
              "start": "0",
              "type": "text",
            },
          ],
          "end": "]",
          "start": "[",
          "type": "square",
        },
        Object {
          "start": ".",
          "type": "dot",
        },
        Object {
          "start": "tokenize",
          "type": "identifiers",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "start": ".",
                  "type": "text",
                },
              ],
              "end": "'",
              "start": "'",
              "type": "string",
            },
          ],
          "end": ")",
          "start": "(",
          "type": "round",
        },
        Object {
          "start": "
        ",
          "type": "text",
        },
        Object {
          "start": "try",
          "type": "keywords",
        },
        Object {
          "start": "
",
          "type": "text",
        },
        Object {
          "children": Array [
            Object {
              "start": "        ",
              "type": "text",
            },
            Object {
              "start": "def",
              "type": "keywords",
            },
            Object {
              "start": "   ",
              "type": "text",
            },
            Object {
              "start": "index",
              "type": "identifiers",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "start": "=",
              "type": "operators",
            },
            Object {
              "start": "     ",
              "type": "text",
            },
            Object {
              "start": "versionMap",
              "type": "identifiers",
            },
            Object {
              "start": ".",
              "type": "dot",
            },
            Object {
              "start": "get",
              "type": "identifiers",
            },
            Object {
              "children": Array [
                Object {
                  "start": "version_type",
                  "type": "identifiers",
                },
              ],
              "end": ")",
              "start": "(",
              "type": "round",
            },
            Object {
              "start": ";",
              "type": "delimiters",
            },
            Object {
              "start": "
",
              "type": "text",
            },
            Object {
              "start": "versionArray",
              "type": "identifiers",
            },
            Object {
              "children": Array [
                Object {
                  "start": "index",
                  "type": "identifiers",
                },
              ],
              "end": "]",
              "start": "[",
              "type": "square",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "start": "=",
              "type": "operators",
            },
            Object {
              "start": "versionArray",
              "type": "identifiers",
            },
            Object {
              "children": Array [
                Object {
                  "start": "index",
                  "type": "identifiers",
                },
              ],
              "end": "]",
              "start": "[",
              "type": "square",
            },
            Object {
              "start": ".",
              "type": "dot",
            },
            Object {
              "start": "toInteger",
              "type": "identifiers",
            },
            Object {
              "end": ")",
              "start": "(",
              "type": "round",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "start": "+",
              "type": "operators",
            },
            Object {
              "start": " 1
",
              "type": "text",
            },
            Object {
              "start": "if",
              "type": "keywords",
            },
            Object {
              "children": Array [
                Object {
                  "start": "   ",
                  "type": "text",
                },
                Object {
                  "start": "reset",
                  "type": "identifiers",
                },
                Object {
                  "start": " ",
                  "type": "text",
                },
              ],
              "end": ")",
              "start": "(",
              "type": "round",
            },
            Object {
              "start": "
",
              "type": "text",
            },
            Object {
              "children": Array [
                Object {
                  "start": "
    ",
                  "type": "text",
                },
                Object {
                  "start": "for",
                  "type": "keywords",
                },
                Object {
                  "children": Array [
                    Object {
                      "start": "int",
                      "type": "keywords",
                    },
                    Object {
                      "start": " ",
                      "type": "text",
                    },
                    Object {
                      "start": "i",
                      "type": "identifiers",
                    },
                    Object {
                      "start": "=",
                      "type": "operators",
                    },
                    Object {
                      "start": "2",
                      "type": "text",
                    },
                    Object {
                      "start": ";",
                      "type": "delimiters",
                    },
                    Object {
                      "start": "i",
                      "type": "identifiers",
                    },
                    Object {
                      "start": ">",
                      "type": "operators",
                    },
                    Object {
                      "start": "index",
                      "type": "identifiers",
                    },
                    Object {
                      "start": ";",
                      "type": "delimiters",
                    },
                    Object {
                      "start": "     ",
                      "type": "text",
                    },
                    Object {
                      "start": "i",
                      "type": "identifiers",
                    },
                    Object {
                      "start": "--",
                      "type": "incdec",
                    },
                  ],
                  "end": ")",
                  "start": "(",
                  "type": "round",
                },
                Object {
                  "start": " ",
                  "type": "text",
                },
                Object {
                  "children": Array [
                    Object {
                      "start": "
        ",
                      "type": "text",
                    },
                    Object {
                      "start": "versionArray",
                      "type": "identifiers",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "start": "i",
                          "type": "identifiers",
                        },
                      ],
                      "end": "]",
                      "start": "[",
                      "type": "square",
                    },
                    Object {
                      "start": "    ",
                      "type": "text",
                    },
                    Object {
                      "start": "=",
                      "type": "operators",
                    },
                    Object {
                      "start": "    0            ",
                      "type": "text",
                    },
                  ],
                  "end": "}",
                  "start": "{",
                  "type": "block",
                },
                Object {
                  "start": "        ",
                  "type": "text",
                },
              ],
              "end": "}",
              "start": "{",
              "type": "block",
            },
            Object {
              "start": "
",
              "type": "text",
            },
          ],
          "end": "}",
          "start": "{",
          "type": "block",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "start": "catch",
          "type": "keywords",
        },
        Object {
          "children": Array [
            Object {
              "start": "       ",
              "type": "text",
            },
            Object {
              "start": "Exception",
              "type": "identifiers",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "start": "e",
              "type": "identifiers",
            },
          ],
          "end": ")",
          "start": "(",
          "type": "round",
        },
        Object {
          "start": " ",
          "type": "text",
        },
        Object {
          "children": Array [
            Object {
              "start": "        ",
              "type": "text",
            },
            Object {
              "start": "println",
              "type": "identifiers",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "start": "Unrecognized version type \\\\\\"version_type\\\\\\" (should be major, minor or patch)",
                      "type": "text",
                    },
                  ],
                  "end": "\\"",
                  "start": "\\"",
                  "type": "string",
                },
              ],
              "end": ")",
              "start": "(",
              "type": "round",
            },
            Object {
              "start": "    ",
              "type": "text",
            },
          ],
          "end": "}",
          "start": "{",
          "type": "block",
        },
        Object {
          "start": "
",
          "type": "text",
        },
        Object {
          "start": "return",
          "type": "keywords",
        },
        Object {
          "start": "             ",
          "type": "text",
        },
        Object {
          "start": "versionArray",
          "type": "identifiers",
        },
        Object {
          "start": ".",
          "type": "dot",
        },
        Object {
          "start": "join",
          "type": "identifiers",
        },
        Object {
          "children": Array [
            Object {
              "start": "                   ",
              "type": "text",
            },
            Object {
              "children": Array [
                Object {
                  "start": ".",
                  "type": "text",
                },
              ],
              "end": "'",
              "start": "'",
              "type": "string",
            },
            Object {
              "start": "                       ",
              "type": "text",
            },
          ],
          "end": ")",
          "start": "(",
          "type": "round",
        },
        Object {
          "start": "
",
          "type": "text",
        },
      ],
      "end": "}",
      "start": "{",
      "type": "block",
    },
    Object {
      "start": "
",
      "type": "text",
    },
    Object {
      "start": "println",
      "type": "identifiers",
    },
    Object {
      "children": Array [
        Object {
          "start": "bumpVersion",
          "type": "identifiers",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "start": "1.2.3",
                  "type": "text",
                },
              ],
              "end": "'",
              "start": "'",
              "type": "string",
            },
            Object {
              "start": ",",
              "type": "delimiters",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "children": Array [
                Object {
                  "start": "minor",
                  "type": "text",
                },
              ],
              "end": "'",
              "start": "'",
              "type": "string",
            },
            Object {
              "start": ",",
              "type": "delimiters",
            },
            Object {
              "start": " ",
              "type": "text",
            },
            Object {
              "start": "true",
              "type": "identifiers",
            },
          ],
          "end": ")",
          "start": "(",
          "type": "round",
        },
      ],
      "end": ")",
      "start": "(",
      "type": "round",
    },
  ],
  "start": "",
  "type": "root",
}
`;

exports[`Parsing and formatting 2`] = `
"def bumpVersion(String target, String version_type, Boolean reset = false) {
    def versionMap = ['major': 0, 'minor': 1, 'patch': 2]
    def versionArray = target.findAll(/d+.d+.d+/)[0].tokenize('.')
    try {
        def index = versionMap.get(version_type);
        versionArray[index] = versionArray[index].toInteger() + 1
        if (reset)
        {
            for (int i = 2; i > index; i--) {
                versionArray[i] = 0
            }
        }
    } catch (Exception e) {
        println(\\"Unrecognized version type \\\\\\"version_type\\\\\\" (should be major, minor or patch)\\")
    }
    return versionArray.join('.')
}
println(bumpVersion('1.2.3', 'minor', true))"
`;
